### PR TITLE
8330027: Identity hashes of archived objects must be based on a reproducible random seed

### PR DESCRIPTION
**Pull request by tstuefe**
CDS archive contains archived objects with identity hashes.

These hashes are deliberately preserved or even generated during dumping. They are generated based on a seed that is initialized randomly on a per-thread basis. These generations precede CDS dump initialization, so they are not affected by the init_random call there, nor would they be affected by [JDK-8323900](https://bugs.openjdk.org/browse/JDK-8323900).

A random seed will not work for dumping archives since it prevents reproducible archive generation. Therefore, when dumping, these seeds must be initiated in a reproducible way.

--- Update

After discussions with Ioi, and several redos, we settled on:

- make sure CDS dump only ever calls ihash generation from one thread. Means, we disable the explicit hash generation we had been doing before, since that one was called from the VM thread, not from the single java thread
- Start out with a constant seed for all threads. This is fine and does not cause collisions between threads, since - see above - we only call ihash generation from a single threads.
- We also assert that we only use a single thread

--- Update Update

The final version I plan to push does not have above mentioned assert, since it turned out to be too tricky and complex to get right, not worth the trouble.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330027](https://bugs.openjdk.org/browse/JDK-8330027): Identity hashes of archived objects must be based on a reproducible random seed (**Bug** - P4)

### Reviewers
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18735/head:pull/18735` \
`$ git checkout pull/18735`

Update a local copy of the PR: \
`$ git checkout pull/18735` \
`$ git pull https://git.openjdk.org/jdk.git pull/18735/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18735`

View PR using the GUI difftool: \
`$ git pr show -t 18735`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18735.diff">https://git.openjdk.org/jdk/pull/18735.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18735#issuecomment-2049663142)

### CODE REVIEW

**Comment by tstuefe** [comment]
Ping @iklam, @calvinccheung

**Comment by rkennke** [comment]
Do we need to generate I-hash on archive generation at all?

**Comment by tstuefe** [comment]
> Do we need to generate I-hash on archive generation at all?

Yes, see https://github.com/openjdk/jdk/blob/087e8e7f3b2d87e6dc3c4e1c49eff7bae88f1e4f/src/hotspot/share/cds/archiveHeapWriter.cpp#L532

**Comment by rkennke** [comment]
> > Do we need to generate I-hash on archive generation at all?
> 
> 
> 
> Yes, see https://github.com/openjdk/jdk/blob/087e8e7f3b2d87e6dc3c4e1c49eff7bae88f1e4f/src/hotspot/share/cds/archiveHeapWriter.cpp#L532

I get that we need to *preserve* I-hashed. But why do we ever have to generate any new ones?

**Comment by tstuefe** [comment]
> > > Do we need to generate I-hash on archive generation at all?
> > 
> > 
> > Yes, see https://github.com/openjdk/jdk/blob/087e8e7f3b2d87e6dc3c4e1c49eff7bae88f1e4f/src/hotspot/share/cds/archiveHeapWriter.cpp#L532
> 
> I get that we need to _preserve_ I-hashed. But why do we ever have to generate any new ones?

That is an excellent question, and here is what I understand:

- The comment states we want to minimize the chance of runtime writes to the archived objects. I assume we do this to prevent COW on that page, such that we have the maximum effect of memory sharing between processes using the same archive.
- But this relies on archive heap region mapping not needing relocation. If we relocate, we touch every object anyway. You only avoid relocation if you manage to map at the same address. 
- The loading code [1] seems to indicate that the CDS heap region is always allocated at the end of the heap (note that we ignore the "preferred_addr" argument). Therefore, we only get the same address iff the heap has the same size and starts at the same address as at dump time. 
- All of that is moot anyway since, with JDK-8294323, we started forcing relocation due to security reasons.

Therefore I think we don't have to actively generate ihashes.

@iklam What do you think?

[1] https://github.com/openjdk/jdk/blob/16061874ffdd1b018fe1cad7e6d8ba8bdbdbbee1/src/hotspot/share/gc/g1/g1CollectedHeap.cpp#L488

**Comment by dcubed-ojdk** [comment]
I fixed a typo in the bug's synopsis. The easiest way to update is with: `/issue JDK-8330027`.

**Comment by tstuefe** [comment]
> I fixed a typo in the bug's synopsis. The easiest way to update is with: `/issue JDK-8330027`.

Thanks Dan :)

/issue JDK-8330027

**Review by calvinccheung — APPROVED** [review]
Looks good.
nit: please update copyright header in thread.hpp.

**Comment by tstuefe** [comment]
@iklam @calvinccheung could you take another look please?

I rewrote this patch to be both minimally invasive and as bulletproof against concurrent activity as possible.

My thoughts on this:
- just initializing the global seed of os::random with a constant makes the ihash vulnerable against concurrent calls to os::random in unrelated threads. At the minimum, it makes us vulnerable against the order of thread start and number of thread starts since each thread constructor did call os::random to init its ihash seed
- My first patch version constified the ihash seed in the Thread constructor, but that still leaves us somewhat vulnerable against the same problem
- This version - also the simplest in my opinion - makes ihash seed generation lazy, on-demand only, the first time a thread generates an ihash. That is the most robust version, since when dumping, only two threads ever generate ihashes - the initial java thread, and the VM thread. Since both run sequentially, not concurrently, the order of ihash generations is deterministic, and since we restrict the seed initialization to those threads that actually do generate ihashes, we can be reasonably safe of getting the same random numbers.

**Comment by iklam** [comment]
> @iklam @calvinccheung could you take another look please?
> 
> I rewrote this patch to be both minimally invasive and as bulletproof against concurrent activity as possible.
> 
> My thoughts on this:
> 
> * just initializing the global seed of os::random with a constant makes the ihash vulnerable against concurrent calls to os::random in unrelated threads. At the minimum, it makes us vulnerable against the order of thread start and number of thread starts since each thread constructor did call os::random to init its ihash seed
> * My first patch version constified the ihash seed in the Thread constructor, but that still leaves us somewhat vulnerable against the same problem
> * This version - also the simplest in my opinion - makes ihash seed generation lazy, on-demand only, the first time a thread generates an ihash. That is the most robust version, since when dumping, only two threads ever generate ihashes - the initial java thread, and the VM thread. Since both run sequentially, not concurrently, the order of ihash generations is deterministic, and since we restrict the seed initialization to those threads that actually do generate ihashes, we can be reasonably safe of getting the same random numbers.

This PR doesn't help CDS in terms of making the contents of archived heap objects deterministic.

The dumping of archived heap objects is very sensitive to (Java thread) context switching: if two concurrent Java threads call `Object.identityHashcode()` on the same object, then the hashcode inside the header of this object will have deterministic values. We cannot easily recover from this with post-processing, as the shapes of the archived hashtables are influenced by the hashcode, and the CDS code doesn't know how to repack the hashtables in Java.

As a result, during `java -Xshare:dump`, we disable the launching of all new Java threads, so there's only a single (main) Java thread running the whole time

https://github.com/openjdk/jdk/blob/235ba9a7025964b1e43149c7102e26b82b2081ad/src/hotspot/share/prims/jvm.cpp#L2948-L2963

Even if we apply this PR, we still cannot run more than one Java thread. Conversely, if we stick to a single Java thread, then the complexity in this PR is not needed.

**Comment by tstuefe** [comment]
> > @iklam @calvinccheung could you take another look please?
> > I rewrote this patch to be both minimally invasive and as bulletproof against concurrent activity as possible.
> > My thoughts on this:
> > 
> > * just initializing the global seed of os::random with a constant makes the ihash vulnerable against concurrent calls to os::random in unrelated threads. At the minimum, it makes us vulnerable against the order of thread start and number of thread starts since each thread constructor did call os::random to init its ihash seed
> > * My first patch version constified the ihash seed in the Thread constructor, but that still leaves us somewhat vulnerable against the same problem
> > * This version - also the simplest in my opinion - makes ihash seed generation lazy, on-demand only, the first time a thread generates an ihash. That is the most robust version, since when dumping, only two threads ever generate ihashes - the initial java thread, and the VM thread. Since both run sequentially, not concurrently, the order of ihash generations is deterministic, and since we restrict the seed initialization to those threads that actually do generate ihashes, we can be reasonably safe of getting the same random numbers.
> 
> This PR doesn't help CDS in terms of making the contents of archived heap objects deterministic.
> 

It does though, demonstratably. Provided one fixes the "os::random seed constant" problem. Once that is fixed, archives are non-deterministic for the reason stated in the PR description. And this PR makes the archive deterministic again. One could argue whether this is the simplest solution (and we do that below), but saying it does not help is just wrong, sorry.

> The dumping of archived heap objects is very sensitive to (Java thread) context switching: if two concurrent Java threads call `Object.identityHashcode()` on the same object, then the hashcode inside the header of this object will have deterministic values. We cannot easily recover from this with post-processing, as the shapes of the archived hashtables are influenced by the hashcode, and the CDS code doesn't know how to repack the hashtables in Java.
> 
> As a result, during `java -Xshare:dump`, we disable the launching of all new Java threads, so there's only a single (main) Java thread running the whole time
> 
> https://github.com/openjdk/jdk/blob/235ba9a7025964b1e43149c7102e26b82b2081ad/src/hotspot/share/prims/jvm.cpp#L2948-L2963
> 
> Even if we apply this PR, we still cannot run more than one Java thread. Conversely, if we stick to a single Java thread, then the complexity in this PR is not needed.

I get all that. 

But the problem is not restricted to the one java thread.

We create ihashes from the single one java thread, as well as the VM thread. They don't run concurrently, so their order of ihash creation calls is fixed for two subsequent runs of the same JVM.

Okay.

We seed each threads ihash RNG in Thread::Thread(). We seed it with `os::random`. `os::random` works with a global seed.

Okay, we can make that global seed constant too. 

But we are betting that *the number of os::random calls happening before the start of the java thread and the VM thread is constant from run to run*. It often is, but that is not guaranteed by any means. If it differs, the VMthread and the one java thread get different seeds, therefore will generate different ihashes.

The order of os::random() calls can change due to multiple reasons:
- any variations in runtime conditions could cause different code paths that cause os::random to be called or not called.
- concurrent threads could call os:;random. There *are* concurrent threads, albeit no java threads.
- Since we call os::random in Thread::Thread(), any fluctuation that changes the number of threads to be started before the VMThread/java thread is started will add to the number of prior os::random calls, thus messing up the RNG sequence.

I get that the chance for this happening is remote, but hunting sources of entropy is frustrating work, and the patch is really very simple. So, why not fix it? I don't share the opinion that this is added complexity.

**Comment by iklam** [comment]
> I get that the chance for this happening is remote, but hunting sources of entropy is frustrating work, and the patch is really very simple. So, why not fix it? I don't share the opinion that this is added complexity.

Why not do it inside `Thread::Thread()`

```
// thread-specific hashCode stream generator state - Marsaglia shift-xor form
  if (CDSConfig::is_dumping_static_archive()) {
     _hashStateX = 0;
  } else {
     _hashStateX = os::random();
  }  
```

**Comment by tstuefe** [comment]
> > I get that the chance for this happening is remote, but hunting sources of entropy is frustrating work, and the patch is really very simple. So, why not fix it? I don't share the opinion that this is added complexity.
> 
> Why not do it inside `Thread::Thread()`
> 
> ```
> // thread-specific hashCode stream generator state - Marsaglia shift-xor form
>   if (CDSConfig::is_dumping_static_archive()) {
>      _hashStateX = 0;
>   } else {
>      _hashStateX = os::random();
>   }  
> ```

Because then it would inject `os::random` into the startup of every thread, not just of every thread that generates iHashes. So it would also fire for GC threads and other thread started before "our" threads. That would make our random sequence vulnerable against order and number of threads started.

**Comment by tstuefe** [comment]
> > > > > I get that the chance for this happening is remote, but hunting sources of entropy is frustrating work, and the patch is really very simple. So, why not fix it? I don't share the opinion that this is added complexity.
> > > > 
> > > > 
> > > > Why not do it inside `Thread::Thread()`
> > > > ```
> > > > // thread-specific hashCode stream generator state - Marsaglia shift-xor form
> > > >   if (CDSConfig::is_dumping_static_archive()) {
> > > >      _hashStateX = 0;
> > > >   } else {
> > > >      _hashStateX = os::random();
> > > >   }  
> > > > ```
> > > 
> > > 
> > > Because then it would inject `os::random` into the startup of every thread, not just of every thread that generates iHashes. So it would also fire for GC threads and other thread started before "our" threads. That would make our random sequence against order and number of threads started.
> > 
> > 
> > My last answer was rubbish, sorry, did not read your comment carefully enough.
> > Yes, your approach would also work, but it would lead to the two threads involved in dumping the archive - VMthread and the one Java thread - using the same seed, hence generating the same sequence of ihashes. That, in turn, can lead to different archived objects carrying the same ihash, which may negatively impact performance later when the archive is used.
> 
> I think it's better to just not compute the identity hash inside the VM thread. Here's what I tried
> 
> [iklam@ad95e2e](https://github.com/iklam/jdk/commit/ad95e2e8b00cb151617463af41648cdece2dfc7b)
> 
> We thought that forcing the identity hash computation would increase sharing across processes, as it would mean fewer updates of the object headers during run time. However, most of the heap objects in the CDS archive are not accessible by the application (they are part of the archived module graph, etc). Also the archive contains a large number of Strings, which are unlikely to need the identity hash (String has its own hashcode() method).
> 
> Since the reason is rather dubious, I think it's better to remove it and simplify the system.

I like that, that is simpler. Okay, then we will only call ihash from a single thread, so a global constant seed should be fine. I should be able to assert that, right?

**Comment by iklam** [comment]
> > I think it's better to just not compute the identity hash inside the VM thread. Here's what I tried
> > [iklam@ad95e2e](https://github.com/iklam/jdk/commit/ad95e2e8b00cb151617463af41648cdece2dfc7b)
> > We thought that forcing the identity hash computation would increase sharing across processes, as it would mean fewer updates of the object headers during run time. However, most of the heap objects in the CDS archive are not accessible by the application (they are part of the archived module graph, etc). Also the archive contains a large number of Strings, which are unlikely to need the identity hash (String has its own hashcode() method).
> > Since the reason is rather dubious, I think it's better to remove it and simplify the system.
> 
> I like that, that is simpler. Okay, then we will only call ihash from a single thread, so a global constant seed should be fine. I should be able to assert that, right?

AI think an assert can be added, since we don't allow any Java threads to be launched. So even test cases that run arbitrary Java code during -Xshare:dump (using Java agents of -XX:ArchiveHeapTestClass) will not be able to run any Java code outside of the main Java thread.

**Comment by tstuefe** [comment]
@iklam Sorry, had to force push because I messed up the PR branch somehow.

The new version contains the change proposed by you - getting rid of explicit ihash generation when dumping - as well as a simplified version of my original patch. We now are back at generating the ihash seed in Thread::Thread(), with a constant if CDS dumps, and on ihash generation we check that we ever only call it with a single thread.

**Comment by iklam** [comment]
> @iklam Sorry, had to force push because I messed up the PR branch somehow.

I think it's possible to just discard your local branch, and check out a new version from the PR, then make changes on top of it. That way you can avoid forced pushes.

> The new version contains the change proposed by you - getting rid of explicit ihash generation when dumping - as well as a simplified version of my original patch. We now are back at generating the ihash seed in Thread::Thread(), with a constant if CDS dumps, and on ihash generation we check that we ever only call it with a single thread.

This version looks good. I am running our tiers 1-4 just to be sure.

Thanks

**Comment by tstuefe** [comment]
> > @iklam Sorry, had to force push because I messed up the PR branch somehow.
> 
> I think it's possible to just discard your local branch, and check out a new version from the PR, then make changes on top of it. That way you can avoid forced pushes.

Its possible, I do this a lot when I mess up locally. But in this case, I had already accidentally pushed a broken merge to my GH fork, and then saw it touches >5000 files and has >300 commits. No idea what went wrong, but the only way to fix this that I found was to locally redo the patch and switch the branch.

> 
> > The new version contains the change proposed by you - getting rid of explicit ihash generation when dumping - as well as a simplified version of my original patch. We now are back at generating the ihash seed in Thread::Thread(), with a constant if CDS dumps, and on ihash generation we check that we ever only call it with a single thread.
> 
> This version looks good. I am running our tiers 1-4 just to be sure.
> 

Cool. I like the new version more, its simpler. Thank you!

> Thanks

**Comment by tstuefe** [comment]
@calvinccheung could you re-review this PR, please? It is simpler than the old version, but completely redone. Thank you!

File: src/hotspot/share/runtime/synchronizer.cpp
Comment (iklam) [inline src/hotspot/share/runtime/synchronizer.cpp:1011]: It should check for `CDSConfig::is_dumping_static_archive()` instead (here and in `Thread::Thread()`), to match the check in [JVM_StartThread]()https://github.com/openjdk/jdk/blob/7c1fad4fb6c387bbfb72b3f96b610e7cbc2ef312/src/hotspot/share/prims/jvm.cpp#L2946-L2948)

File: src/hotspot/share/runtime/synchronizer.cpp
Comment (iklam) [inline src/hotspot/share/runtime/synchronizer.cpp:952]: I think this is not thread safe, as I got few intermittent asserts.

**Review by iklam — CHANGES_REQUESTED** [review]

File: src/hotspot/share/runtime/synchronizer.cpp
Comment (tstuefe) [reply src/hotspot/share/runtime/synchronizer.cpp:952]: Hmm, it should not really need to be thread safe, since it should only ever be called from a single thread. Do the asserts always happen in the one and only java thread?

Comment (iklam) [reply src/hotspot/share/runtime/synchronizer.cpp:952]: I found the problem. A new `JavaThread` is created during VM exit. This thread is attached to the main OS thread:

```
Thread 2 "java" hit Breakpoint 3, JavaThread::JavaThread (this=0x7ffff002c4e0, is_attaching_via_jni=true) at javaThread.cpp:532
532	JavaThread::JavaThread(bool is_attaching_via_jni) : JavaThread() {
(gdb) where
#0  JavaThread::JavaThread (this=0x7ffff002c4e0, is_attaching_via_jni=true)
#1  attach_current_thread (vm=0x7ffff7c283d8 <main_vm>, penv=0x7ffff5358d50, _args=0x7ffff5358d60, daemon=false)
#2  jni_AttachCurrentThread (vm=0x7ffff7c283d8 <main_vm>, penv=0x7ffff5358d50, _args=0x7ffff5358d60)
#3  JavaVM_::AttachCurrentThread (this=0x7ffff7c283d8 <main_vm>, penv=0x7ffff5358d50, args=0x7ffff5358d60)
#4  jni_DestroyJavaVM_inner (vm=0x7ffff7c283d8 <main_vm>)
#5  jni_DestroyJavaVM (vm=0x7ffff7c283d8 <main_vm>)
#6  JavaMain ()
#7  ThreadJavaMain ()
#8  start_thread (arg=<optimized out>)
#9  clone3 ()
```

This thread can execute Java code (shutdown hooks) which can call `Object::identityHashcode()`.

**Comment by tstuefe** [comment]
@iklam I decided to remove the assert. I briefly thought about adding a condition to time (don't assert after dump finished) but decided against more complexity.

**Review by iklam — APPROVED** [review]
LGTM.

**Comment by tstuefe** [comment]
I'll wait for @calvinccheung 's re-approval, then push.

**Review by calvinccheung — APPROVED** [review]
The latest version looks good. Thanks!

**Comment by tstuefe** [comment]
Cool, thanks @calvinccheung !

/integrate

**Pull request closed without merging**
